pylops_mpi/basicoperators/FirstDerivative.py

```diff
@@ -0,0 +1,285 @@
+    creating additional **"ghost cells"** around the boundary of each process's local data domain. These ghost cells
+    are replicated copies of the border cells from neighboring processes, these cells allow each process
+    to perform local computations without the need for explicit communication with other processes.
+
+    Now, for a one-dimensional DistributedArray, the first order forward stencil is:
+
+    .. math::
+        y[i] = (x[i+1] - x[i]) / \Delta x
+
+    while the first order backward stencil is:
+
+    .. math::
+        y[i] = (x[i] - x[i-1]) / \Delta x
+
+    and the second-order centered stencil is:
+
+    .. math::
+        y[i] = 0.5 * (x[i+1] - x[i-1]) / \Delta x
+
+    where :math:`y` is a DistributedArray and :math:`x` is a Ghosted array created by adding
+    border cells of neighboring processes to the local array at each rank.
+
+    Formulas for the third-order centered stencil can be found at
+    this `link <https://en.wikipedia.org/wiki/Finite_difference_coefficient>`_.
+
+    """
+
+    def __init__(self,
+                 dims: int,
+                 sampling: float = 1.0,
+                 kind: str = "centered",
+                 edge: bool = False,
+                 order: int = 3,
+                 dtype: DTypeLike = np.float64):
+        self.dims = dims
+        shape = (dims, dims)
+        super().__init__(shape=shape, dtype=np.dtype(dtype))
+        self.sampling = sampling
+        self.kind = kind
+        self.edge = edge
+        self.order = order
+        self._register_multiplications(self.kind, self.order)
+
+    def _register_multiplications(
+            self,
+            kind: str,
+            order: int,
+    ) -> None:
+        # choose _matvec and _rmatvec kind
+        self._hmatvec: Callable
+        self._hrmatvec: Callable
+        if kind == "forward":
+            self._hmatvec = self._matvec_forward
+            self._hrmatvec = self._rmatvec_forward
+        elif kind == "centered":
+            if order == 3:
+                self._hmatvec = self._matvec_centered3
+                self._hrmatvec = self._rmatvec_centered3
+            elif order == 5:
+                self._hmatvec = self._matvec_centered5
+                self._hrmatvec = self._rmatvec_centered5
+            else:
+                raise NotImplementedError("'order' must be '3, or '5'")
+        elif kind == "backward":
+            self._hmatvec = self._matvec_backward
+            self._hrmatvec = self._rmatvec_backward
+        else:
+            raise NotImplementedError(
+                "'kind' must be 'forward', 'centered', or 'backward'"
+            )
+
+    def _matvec(self, x: DistributedArray) -> DistributedArray:
+        # If Partition.BROADCAST, then convert to Partition.SCATTER
+        if x.partition is Partition.BROADCAST:
+            x = DistributedArray.to_dist(x=x.local_array, axis=x.axis)
```

Comment:
mmh if the array is of `Partition.BROADCAST`, what does `x.axis` really mean? I guess from how DistributedArray is implemented right now, one could choose any axis, but here this would lead to problems... see my general comment